staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go

```diff
@@ -213,29 +213,96 @@ func (s *SpdyRoundTripper) proxyAuth(proxyURL *url.URL) string {
 // clients may call SpdyRoundTripper.Connection() to retrieve the upgraded
 // connection.
 func (s *SpdyRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	// TODO what's the best way to clone the request?
-	r := *req
-	req = &r
-	req.Header.Add(httpstream.HeaderConnection, httpstream.HeaderUpgrade)
-	req.Header.Add(httpstream.HeaderUpgrade, HeaderSpdy31)
+	const (
+		maxRedirects = 10
+	)
 
-	conn, err := s.dial(req)
-	if err != nil {
-		return nil, err
-	}
+	var (
+		location         = req.URL
+		method           = req.Method
+		intermediateConn net.Conn
+		resp             *http.Response
+	)
 
-	err = req.Write(conn)
-	if err != nil {
-		return nil, err
+	defer func() {
+		if intermediateConn != nil {
+			intermediateConn.Close()
+		}
+	}()
+
+	r := new(http.Request)
+	// shallow clone
+	*r = *req
+	// deep copy headers
+	header := make(http.Header, len(req.Header))
+	for key, values := range req.Header {
+		newValues := make([]string, len(values))
+		copy(newValues, values)
+		header[key] = newValues
 	}
 
-	resp, err := http.ReadResponse(bufio.NewReader(conn), req)
-	if err != nil {
-		return nil, err
-	}
+	header.Add(httpstream.HeaderConnection, httpstream.HeaderUpgrade)
+	header.Add(httpstream.HeaderUpgrade, HeaderSpdy31)
+
+redirectLoop:
```

Comment:
This duplicates a lot of code from the proxy (https://github.com/kubernetes/kubernetes/blob/c3463d737efe553f4818ff6495f93856eac4557b/staging/src/k8s.io/apiserver/pkg/registry/generic/rest/proxy.go#L265). Is it possible to extract a shared method? Maybe something that takes a `connectBackend` function?